# Exercise 04 - Due Friday, October 21 at 12pm

*Objectives*: Gain experience loading a CSV dataset into a database model you define yourself and using SQL to explore its contents. Explore the data by writing and executing a number of SQL queries using common syntax and functions and describing your findings.

*Grading criteria*: The tasks should all be completed, and questions should all be answered with SQL queries in the space provided, unless a text answer is requested. Results should be correct, according to the embedded tests. The notebook itself should be completely reproducible; from start to finish, another person should be able to use the same code to obtain the same results as yours.  Note that you will receive no more than partial credit if you do not add text/markdown cells explaining your thinking at each major step in each problem.

*Deadline*: Friday, October 21, 12pm. Submit your notebook to Blackboard and push it to your GitHub repository.

## Setup - obtain data and create database

The US City Open Data Census has data on [restaurant inspections from all over the country](http://us-city.census.okfn.org/dataset/food-safety).  Let's take a look at data from Albuquerque.  *Note*: metadata and other details for this dataset are available at http://data.cabq.gov/business/foodinspections/.

Fyi, to save you a step, I've converted this file into a clean CSV file by using the `dos2unix` command to change its line endings and to switch it to use commas instead of tabs using `csvformat -t inputfile.csv > outputfile.csv`.

First we download the dataset:

In [1]:
!wget https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/exercises/abq-food-inspections-fy2010.csv

--2016-10-24 13:59:38--  https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/exercises/abq-food-inspections-fy2010.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.32.133
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.32.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 5025226 (4.8M) [text/plain]
Saving to: ‘abq-food-inspections-fy2010.csv’

abq-food-inspection 100%[=====================>]   4.79M  --.-KB/s   in 0.1s   

2016-10-24 13:59:38 (35.4 MB/s) - ‘abq-food-inspections-fy2010.csv’ saved [5025226/5025226]



## Problem 1 - examine the data (35 points)

Use `csvstat` and other tools as you see fit and observe its contents.  Review the metadata definitions linked to above.  Describe what you see:  What do all the columns mean?  Are there null values?  Which columns interest you the most?  Which columns present some complications or questions you would like to ask?

Basic Inspection of the data: 

In [2]:
!csvstat abq-food-inspections-fy2010.csv

  1. FACILITY_NAME
	<class 'str'>
	Nulls: False
	Unique values: 2615
	5 most frequent values:
		SMITHS FOOD AND DRUG CENTERS INC:	154
		ISOTOPES PARK - OVATIONS:	93
		PER EHSII CECELIA GARCIA OOB LA REYNA DEL SUR:	85
		ECHO INC:	68
		ISLETA AMPHITHEATER:	68
	Max length: 77
  2. FACILITY_KEY
	<class 'int'>
	Nulls: False
	Min: 13
	Max: 101482
	Sum: 820405773
	Mean: 46730.79135338346
	Median: 50866.0
	Standard Deviation: 32659.794578837962
	Unique values: 2835
	5 most frequent values:
		46323:	93
		65863:	85
		80949:	68
		32817:	68
		91825:	66
  3. SITE_ADDRESS
	<class 'str'>
	Nulls: True
	Unique values: 2456
	5 most frequent values:
		1601 AVENIDA CESAR CHAVEZ SE:	94
		644 OLD COORS BLVD SW:	85
		8601 CENTRAL AV NE:	77
		2200 SUNPORT BLVD SE:	75
		5210 CENTRAL AV SE:	73
	Max length: 35
  4. CITY
	<class 'str'>
	Nulls: False
	Values: ALBUQUERQUE
  5. STATE
	<class 'str'>
	Nulls: False
	Values: NM, 87
  6. ZIP
	<class 'str'>
	Nulls: False
	Unique values: 48
	5 most frequent values:
		87110

What do all the colums mean?

Column 1: FACILITY NAME
It is a catagorical variable that has 2615 unique values. This is the name of facilities that was inspected during the  2010 fiscal year in Albuquerque. 

Column 2: FACILITY KEY
It is a numrical variable. However, facility key correspond to the respective facility name. Therefore should also be considered as catagorical variable. 

Column 3: SITE ADDRESS; Column 4: CITY; Column 5: STATE; Column 6: Zip
Categorical variable. The address of each resturant inspected. 

Column 7, 8: OWNER_KEY, OWNDER_NAME
the name and identifier of the owner of each resturant being inspected. 

Column 9: NATURE_OF_BUSINESS
The nature of business,catagorical variable, what type of business does one resturant run. 

Column 10,11,12: STREET_NUMBER, STREET_NAME, STREET_TYPE
Breaks up column 3 to the number in the street address, street name and the type of street and put them in three seperate columns 

Column 13: POST_DIRECTIONAL
the section of the city the resturant is in. 

Column 14: PHONE
phone number of each resturant. 

Column 15,16: PROGRAM_CATEGORY,PROGRAM_CATEGORY_DESCRIPTION
Categorical variables, eachPROGRAM_CATEGORY_DESCRIPTION describes what the resturant function as. For example, there are food processor or food service establishment. 

Column 17: INSPECTION_DATE
The date of inspection

Column 18: INSPECTION_TYPE
All null values. The type of inspection that shouldbe performed. 

Column 19:INSPECTION_DESC
The content of the inspection at that time. such as foutine food inspection, emergency response, new business
inspection, etc. 

Column 20: SERIAL_NUM
Serial number associated with each inspection. 

Column 21,22:ACTION_CODE, ACTION_DESC
The action that the inspector took with each inspection at the corresponding time and resturant. Each ACTION_DESC has a corresponding unique ACTION_CODE

Column 23,24:RESULT_CODE, RESULT_DESC
The result of the inspection at the corresponding time and resturant. Each RESULT_DESC has a corresponding unique RESULT_CODE

Column 25,26:VIOLATION_CODE, VIOLATION_DESC
The violation that the resturant made at that time. Each VIOLATION_DESC has a corresponding unique VIOLATION_CODE

Column 27: INSPECTION_MEMO
Any memerizable details of each inspection. 

Are there null values?

Yes. Column 14: PHONE, Column 9: NATURE_OF_BUSINESS, Column 18: INSPECTION_TYPE , Column 25,26:VIOLATION_CODE, and VIOLATION_DESC all contain null values. In fact, Column 18: INSPECTION_TYPE only contains null values. 

Which columns interest you the most? 

Column 15: PROGRAM_CATEGORY and Column 16: PROGRAM_CATEGORY_DESCRIPTION is interesting because the key variable PROGRAM_CATEGORY doesn't have null values but the Column 16: PROGRAM_CATEGORY_DESCRIPTION has. 
for the purpose of the dataset, the violation columns interest me the most because I assume for this data set we are to find out which rule was violated the most and which resturants are the most likely to commit violations. 

Which columns present some complications or questions you would like to ask? 

I don't see any columns present complication. 

## Problem 2 - define a database model and load the data (35 points)

Based on what you found above, create and connect to a new database, define a database table in it, and load this dataset into it.  You may use either of the methods for this step you have seen in earlier class notebooks.  You may choose to eliminate variables/columns if they are not relevant or interesting to you - explain your reasoning if you do.  Either way, you should load a majority of the columns present in the source dataset into the database and all of its rows.

Once your data has loaded successfully, run a `COUNT(*)` query to verify that all the data has loaded correctly.

Insert a combination of code and text/markdown cells here to connect to your database, define your table, load its data, and verify the loaded records.

In [3]:
%load_ext sql

/opt/conda/lib/python3.5/site-packages/IPython/config.py:13: ShimWarning: The `IPython.config` package has been deprecated. You should import from traitlets.config instead.
  "You should import from traitlets.config instead.", ShimWarning)
/opt/conda/lib/python3.5/site-packages/IPython/utils/traitlets.py:5: UserWarning: IPython.utils.traitlets has moved to a top-level traitlets package.
  warn("IPython.utils.traitlets has moved to a top-level traitlets package.")


In [4]:
!echo 'redspot' | sudo -S service postgresql restart

[sudo] password for jovyan: Restarting PostgreSQL 9.5 database server: main.


In [5]:
!createdb -U dbuser exercise4

In [6]:
%sql postgresql://dbuser@localhost:5432/exercise4

'Connected: dbuser@exercise4'

In [7]:
%%sql
DROP TABLE IF EXISTS foodinsp;
CREATE TABLE foodinsp 
(
    FACILITY_NAME CHAR(202),
    FACILITY_KEY INTEGER,
    SITE_ADDRESS CHAR(102),
    CITY CHAR(52),
    STATE CHAR(6),
    ZIP CHAR(22),
    OWNER_KEY INTEGER,
    OWNER_NAME CHAR(202),
    NATURE_OF_BUSINESS CHAR(25),
    STREET_NUMBER  CHAR(16),
    STREET_NAME CHAR(52),
    STREET_TYPE CHAR(10),
    POST_DIRECTIONAL CHAR(6),
    PHONE CHAR(42),
    PROGRAM_CATEGORY CHAR(10),
    PROGRAM_CATEGORY_DESCRIPTION CHAR(92),
    INSPECTION_DATE TIMESTAMP,
    INSPECTION_TYPE CHAR(4),
    INSPECTION_DESC CHAR(102),
    SERIAL_NUM CHAR(20),
    ACTION_CODE CHAR(6),
    ACTION_DESC CHAR(102),
    RESULT_CODE CHAR(6),
    RESULT_DESC CHAR(102),
    VIOLATION_CODE CHAR(14),
    VIOLATION_DESC VARCHAR(202),
    INSPECTION_MEMO CHAR(502)
)

Done.
Done.


[]

In [8]:
!pwd

/home/jovyan/work


In [9]:
%%sql
COPY foodinsp FROM '/home/jovyan/work/abq-food-inspections-fy2010.csv'
CSV
HEADER
QUOTE '"'
DELIMITER ',';

17556 rows affected.


[]

In [10]:
%%sql
SELECT COUNT(*) FROM foodinsp;

1 rows affected.


[(17556,)]

## Problem 3 - explore your data (30 points)

Now that the data is loaded, ask some questions of it!  Identify key variables of interest and note their ranges along with other useful descriptive statistics.  Choose and define a few lines of inquiry, execute queries for each, and describe what you find in your result sets.  Use any query techniques we've seen in class, including aggregate functions, transformations, subqueries, or others as appropriate. 

If you find interesting patterns, adding plots to your exploration would be useful.

Insert a combination of code and text/markdown cells here to explore your data.

① What's the five most common reasons why a resturant is inspected?

In [11]:
%%sql
SELECT inspection_desc, COUNT(*) AS count
FROM foodinsp
GROUP BY inspection_desc
ORDER BY count DESC
LIMIT  5

5 rows affected.


[('ROUTINE FOOD INSPECTION                                                                               ', 12439),
 ('EMERGENCY RESPONSE                                                                                    ', 1987),
 ('New Business Inspection                                                                               ', 1005),
 ('PRE-OPENING FOOD INSPECTION                                                                           ', 743),
 ('ADDITIONAL SERVICE PROVIDED                                                                           ', 439)]

We can see that aside from rountine food inspection, the most common reason for a inspection is emergency response and new business inspection. A reasonable next project would be finding out what those emergency response consist of and try to eliminate them. 

② Which are the facilities that gets the most inspection from authority? This information is quite usful when the authority trys to do planning or use precautions on. 

In [12]:
%%sql
SELECT facility_name, COUNT(*) AS count
FROM foodinsp
WHERE inspection_desc IN (
    SELECT inspection_desc
    FROM foodinsp
    GROUP BY inspection_desc
    ORDER BY COUNT(*) DESC
    LIMIT 1)
GROUP BY facility_name
ORDER BY count DESC
LIMIT 5;

5 rows affected.


[('SMITHS FOOD AND DRUG CENTERS INC                                                                                                                                                                          ', 130),
 ('ISOTOPES PARK - OVATIONS                                                                                                                                                                                  ', 93),
 ('ISLETA AMPHITHEATER                                                                                                                                                                                       ', 68),
 ('ECHO INC                                                                                                                                                                                                  ', 65),
 ('OOB  PROS RANCH MARKETS OF ALBUQUERQUE                                                                                                                                                                    ', 52)]

The Smiths Food and drug centers INC got 130 inspections in the 2010 fiscal year! What is going on there? They had significantly higher visits from authority than any other facilities. Let's find out for what reason they were inspected so much. 

③ The reason to visit Smiths Food and Drug Centers. 

In [13]:
%%sql
SELECT INSPECTION_DESC,COUNT(*) As count
FROM foodinsp
WHERE inspection_desc IN (
    SELECT inspection_desc
    FROM foodinsp
    GROUP BY inspection_desc
    ORDER BY COUNT(*) DESC)
    AND FACILITY_NAME = 'SMITHS FOOD AND DRUG CENTERS INC'
GROUP BY inspection_desc
ORDER BY count DESC
;

4 rows affected.


[('ROUTINE FOOD INSPECTION                                                                               ', 130),
 ('EMERGENCY RESPONSE                                                                                    ', 15),
 ('PHONE CALL                                                                                            ', 7),
 ('ADDITIONAL SERVICE PROVIDED                                                                           ', 2)]

Interesting, most of those checks are routine food inspection. My guess is that this is a huge franchise under the same name. Different from 711 in the area, all the Smiths Food and Drug Centers have used the same name. 

④ Is Smiths Food and Drug Centers using the same name for all its franchise? 

In [14]:
%%sql
SELECT FACILITY_KEY,COUNT(*) As count
FROM foodinsp
WHERE FACILITY_KEY IN (
    SELECT FACILITY_KEY
    FROM foodinsp
    GROUP BY FACILITY_KEY
    ORDER BY COUNT(*) DESC)
    AND FACILITY_NAME = 'SMITHS FOOD AND DRUG CENTERS INC'
GROUP BY FACILITY_KEY
ORDER BY count DESC
;

7 rows affected.


[(4298, 26),
 (4296, 25),
 (4293, 24),
 (4294, 22),
 (4292, 21),
 (4297, 18),
 (4295, 18)]

Since facility_key is unique to each resturant at its unique location, we use that to identify each resturant. Now that's a reasonable count for each resturant. We as researchers don't have to worry about Smiths Food and Drug Centers being a hub for food hygene problems. 

⑤ Let's find out the most violated rule in Alburqurque. 

In [15]:
%%sql
SELECT violation_desc, COUNT(*) AS count
FROM foodinsp
WHERE inspection_desc IN (
    SELECT inspection_desc
    FROM foodinsp
    GROUP BY inspection_desc
    ORDER BY COUNT(*) DESC
    LIMIT 1)
GROUP BY violation_desc
ORDER BY count DESC
LIMIT 5;

5 rows affected.


[('Additional Comments', 5052),
 ('No Violations Found', 3048),
 ('Plumbing: Improperly installed/maintained/supplied', 445),
 ('Non-Food contact surfaces', 407),
 ('Improper sanitizing', 398)]

Gross! the most violation aside from additional comments and no violatations found is plumbing problem.

⑥ Who has the most series plumbing issue through out the year?

In [16]:
%%sql
SELECT FACILITY_NAME,COUNT(*) As count
FROM foodinsp
WHERE FACILITY_NAME IN (
    SELECT FACILITY_NAME
    FROM foodinsp
    GROUP BY FACILITY_NAME
    ORDER BY COUNT(*) DESC)
    AND violation_desc = 'Plumbing: Improperly installed/maintained/supplied'
GROUP BY FACILITY_NAME
ORDER BY count DESC
;

408 rows affected.


[('EL MEZQUITE MARKET LLC                                                                                                                                                                                    ', 4),
 ('99 BANH SUPERMARKET                                                                                                                                                                                       ', 4),
 ('SMITHS FOOD AND DRUG CENTERS INC                                                                                                                                                                          ', 4),
 ('MCDONALDS                                                                                                                                                                                                 ', 4),
 ('KEVA JUICE SW LLC                                                                                                                                                                                         ', 3),
 ('ALBERTSONS 904                                                                                                                                                                                            ', 3),
 ('UPTOWN SPORTS GRILL                                                                                                                                                                                       ', 3),
 ('OOB WILEYS PLACE                                                                                                                                                                                          ', 3),
 ('EL MEZQUITE MARKET                                                                                                                                                                                        ', 3),
 ('GRAHAM CENTRAL STATION                                                                                                                                                                                    ', 3),
 ('OOB NOTHING BUT NOODLES                                                                                                                                                                                   ', 3),
 ('DUDS N SUDS                                                                                                                                                                                               ', 3),
 ('HAWTHORN INN AND SUITES                                                                                                                                                                                   ', 2),
 ('WINNING COFFEE CO                                                                                                                                                                                         ', 2),
 ('OOB PUDGE BROS PIZZA                                                                                                                                                                                      ', 2),
 ('OOB BIRD OF PARADISE/JACKS PIZZA                                                                                                                                                                          ', 2),
 ('BRAVO CUCINA ITALIANA                                                                                                                                                                                     ', 2),
 ('CHICADEES                                                                                                                                                                                                 ', 2),
 ('OUEIS GAS                                                                                                                                                                            

We can see that EL MEZQUITE MARKET LLC, 99 BANH SUPERMARKET, SMITHS FOOD AND DRUG CENTERS INC, MCDONALDS are the resturants with most plumbing related issues. We know that SMITHS FOOD AND DRUG CENTERS INC and  MCDONALDS are big chain stores that contain a series of resturants. So this may not be as helpful. So who has most plumbing problem in fy 2010?

⑦ the unique resturant with most plumbing problem. 

In [17]:
%%sql
SELECT FACILITY_KEY, COUNT(*) As count
FROM foodinsp
WHERE FACILITY_KEY IN (
    SELECT FACILITY_KEY
    FROM foodinsp
    GROUP BY FACILITY_KEY
    ORDER BY COUNT(*) DESC)
    AND violation_desc = 'Plumbing: Improperly installed/maintained/supplied'
GROUP BY FACILITY_KEY
ORDER BY count DESC
;

422 rows affected.


[(18468, 4),
 (60817, 4),
 (58022, 3),
 (42272, 3),
 (80654, 3),
 (33793, 3),
 (3066, 3),
 (38941, 3),
 (1481, 2),
 (7604, 2),
 (12414, 2),
 (4296, 2),
 (12288, 2),
 (24079, 2),
 (153, 2),
 (79955, 2),
 (47776, 2),
 (8499, 2),
 (75923, 2),
 (11963, 2),
 (15225, 2),
 (87552, 2),
 (51074, 2),
 (3069, 2),
 (72855, 2),
 (4291, 2),
 (58961, 2),
 (39922, 2),
 (14547, 2),
 (79228, 2),
 (15018, 2),
 (4917, 2),
 (86830, 2),
 (2924, 2),
 (32413, 2),
 (40040, 2),
 (83477, 2),
 (80880, 2),
 (933, 2),
 (7184, 2),
 (87362, 2),
 (69963, 2),
 (2804, 2),
 (77110, 2),
 (81446, 2),
 (78083, 2),
 (67831, 2),
 (82998, 2),
 (24006, 2),
 (30715, 2),
 (44747, 2),
 (70210, 2),
 (71747, 2),
 (5599, 2),
 (76698, 2),
 (2140, 2),
 (3068, 2),
 (48566, 2),
 (5870, 2),
 (3496, 2),
 (46358, 1),
 (9528, 1),
 (2682, 1),
 (81278, 1),
 (2082, 1),
 (32412, 1),
 (5727, 1),
 (12297, 1),
 (39172, 1),
 (90394, 1),
 (21374, 1),
 (11447, 1),
 (58867, 1),
 (23036, 1),
 (33404, 1),
 (85156, 1),
 (88281, 1),
 (4749, 1),
 (20311, 1),
 (44457, 1),
 (72205, 1),
 (60406, 1),
 (84282, 1),
 (80227, 1),
 (40547, 1),
 (35675, 1),
 (19535, 1),
 (72003, 1),
 (82916, 1),
 (5036, 1),
 (434, 1),
 (69594, 1),
 (717, 1),
 (46360, 1),
 (2107, 1),
 (34318, 1),
 (91796, 1),
 (16253, 1),
 (78479, 1),
 (83150, 1),
 (22730, 1),
 (2579, 1),
 (60797, 1),
 (73675, 1),
 (45230, 1),
 (25529, 1),
 (16247, 1),
 (47417, 1),
 (73169, 1),
 (17689, 1),
 (39784, 1),
 (78763, 1),
 (58875, 1),
 (75567, 1),
 (3520, 1),
 (87156, 1),
 (63998, 1),
 (16344, 1),
 (86678, 1),
 (43835, 1),
 (8178, 1),
 (90071, 1),
 (8336, 1),
 (9562, 1),
 (10261, 1),
 (21456, 1),
 (65939, 1),
 (59138, 1),
 (90924, 1),
 (18550, 1),
 (2195, 1),
 (8321, 1),
 (3116, 1),
 (73366, 1),
 (12417, 1),
 (21363, 1),
 (19065, 1),
 (78160, 1),
 (73230, 1),
 (18888, 1),
 (25071, 1),
 (83808, 1),
 (70832, 1),
 (4299, 1),
 (5963, 1),
 (157, 1),
 (86849, 1),
 (81030, 1),
 (79570, 1),
 (34264, 1),
 (46323, 1),
 (69384, 1),
 (89556, 1),
 (87091, 1),
 (2298, 1),
 (7311, 1),
 (83906, 1),
 (5899, 1),
 (5872, 1),
 (86793, 1),
 (17770, 1),
 (2588, 1),
 (75997, 1),
 (78575, 1),
 (15134, 1),
 (92610, 1),
 (76060, 1),
 (70008, 1),
 (9615, 1),
 (85593, 1),
 (72152, 1),
 (75677, 1),
 (84328, 1),
 (43473, 1),
 (89243, 1),
 (88336, 1),
 (65920, 1),
 (87096, 1),
 (5500, 1),
 (4300, 1),
 (73713, 1),
 (3994, 1),
 (90175, 1),
 (83439, 1),
 (36509, 1),
 (68677, 1),
 (60077, 1),
 (87359, 1),
 (86770, 1),
 (45451, 1),
 (89405, 1),
 (11510, 1),
 (92649, 1),
 (65639, 1),
 (87734, 1),
 (87772, 1),
 (4292, 1),
 (15081, 1),
 (7601, 1),
 (21354, 1),
 (84295, 1),
 (80911, 1),
 (86912, 1),
 (4754, 1),
 (66045, 1),
 (41235, 1),
 (12054, 1),
 (32714, 1),
 (1900, 1),
 (46359, 1),
 (75684, 1),
 (77229, 1),
 (16751, 1),
 (39943, 1),
 (6639, 1),
 (25200, 1),
 (17811, 1),
 (70581, 1),
 (43480, 1),
 (56117, 1),
 (8949, 1),
 (84607, 1),
 (7175, 1),
 (67719, 1),
 (85601, 1),
 (31322, 1),
 (70112, 1),
 (86655, 1),
 (12992, 1),
 (64816, 1),
 (85368, 1),
 (75079, 1),
 (4290, 1),
 (2802, 1),
 (81277, 1),
 (84462, 1),
 (43860, 1),
 (93519, 1),
 (85894, 1),
 (63958, 1),
 (74245, 1),
 (77098, 1),
 (2149, 1),
 (42280, 1),
 (12014, 1),
 (85891, 1),
 (2154, 1),
 (26429, 1),
 (2128, 1),
 (65679, 1),
 (90646, 1),
 (68667, 1),
 (85797, 1),
 (85598, 1),
 (91853, 1),
 (27985, 1),
 (10087, 1),
 (16261, 1),
 (4288, 1),
 (67282, 1),
 (3492, 1),
 (1902, 1),
 (7017, 1),
 (43546, 1),
 (72434, 1),
 (3490, 1),
 (21407, 1),
 (7257, 1),
 (55986, 1),
 (80110, 1),
 (20511, 1),
 (35062, 1),
 (67291, 1),
 (69630, 1),
 (76720, 1),
 (14951, 1),
 (86029, 1),
 (10089, 1),
 (67825, 1),
 (83186, 1),
 (60732, 1),
 (85595, 1),
 (51985, 1),
 (8627, 1),
 (28313, 1),
 (89033, 1),
 (49615, 1),
 (75360, 1),
 (76808, 1),
 (8925, 1),
 (44321, 1),
 (69571, 1),
 (80199, 1),
 (62785, 1),
 (21356, 1),
 (6502, 1),
 (86637, 1),
 (43447, 1),
 (85964, 1),
 (21346, 1),
 (67251, 1),
 (75805, 1),
 (84179, 1),
 (46797, 1),
 (80949, 1),
 (60249, 1),
 (23121, 1),
 (83767, 1),
 (81740, 1),
 (87919, 1),
 (2141, 1),
 (5869, 1),
 (40834, 1),
 (8082

Then who is 18468 and 60817? 

In [18]:
%%sql
SELECT FACILITY_KEY, FACILITY_NAME
FROM foodinsp
WHERE FACILITY_KEY = '18468' 
OR FACILITY_KEY = '60817'

93 rows affected.


[(18468, '99 BANH SUPERMARKET                                                                                                                                                                                       '),
 (18468, '99 BANH SUPERMARKET                                                                                                                                                                                       '),
 (18468, '99 BANH SUPERMARKET                                                                                                                                                                                       '),
 (18468, '99 BANH SUPERMARKET                                                                                                                                                                                       '),
 (18468, '99 BANH SUPERMARKET                                                                                                                                                                                       '),
 (18468, '99 BANH SUPERMARKET                                                                                                                                                                                       '),
 (18468, '99 BANH SUPERMARKET                                                                                                                                                                                       '),
 (18468, '99 BANH SUPERMARKET                                                                                                                                                                                       '),
 (18468, '99 BANH SUPERMARKET                                                                                                                                                                                       '),
 (18468, '99 BANH SUPERMARKET                                                                                                                                                                                       '),
 (18468, '99 BANH SUPERMARKET                                                                                                                                                                                       '),
 (18468, '99 BANH SUPERMARKET                                                                                                                                                                                       '),
 (18468, '99 BANH SUPERMARKET                                                                                                                                                                                       '),
 (18468, '99 BANH SUPERMARKET                                                                                                                                                                                       '),
 (18468, '99 BANH SUPERMARKET                                                                                                                                                                                       '),
 (18468, '99 BANH SUPERMARKET                                                                                                                                                                                       '),
 (18468, '99 BANH SUPERMARKET                                                                                                                                                                                       '),
 (18468, '99 BANH SUPERMARKET                                                                                                                                                                                       '),
 (18468, '99 BANH SUPERMARKET                                                                                   

99 BANH SUPERMARKET and EL MEZQUITE MARKET LLC should work on their plumbing problem next year. 

## Bonus (10 points)

We have seen a few simple techniques for normalizing data in class.  Choose a column or set of columns that might be redundant and normalize them out to another table:

* Identify the columns to normalize in the source data table
* Create one or more new tables to hold the normalized records
* Extract data into the new tables
* Re-insert the new foreign key references from the new table into the source table